### PR TITLE
Expose dBFS when doing audio analysis

### PR DESCRIPTION
**Pull request by jpverdejo**
This PR publishes the audio volume (in [dBFS](https://en.wikipedia.org/wiki/DBFS)) via the configured publishers.

The intention is to use this metadata to trigger automations (low or high thresholds, meaning that something is louder or quieter than expected) and we could even add a sound level indicator to the lovelace cards like baby monitors do

image

This is just a draft and I'm sure I'm breaking patterns here, so happy to improve based on your feedback

### CODE REVIEW

**Comment by NickM-27** [comment]
I know it's a draft but just wanted to add that I don't think http is the right way to go for publishing to mqtt. I think it would be better to have another type of comms class which uses the faster_fifo Queue to send messages across processes

**Comment by jpverdejo** [comment]
@NickM-27 That makes sense. Just to be clear you're proposing something like:

/comms/publisher.py
```
class publisher:
  init(dispatcher):
    self.queue = Queue()
    self.dispatcher = dispatcher
  publish(msg):
    add_to_queue(msg)
  run:
    queue_reader_loop:
        self.dispatcher.publish(...)
```

/app.py
```
publisher = Publisher()
audio_process = mp.Process(
                target=listen_to_audio,
                name="audio_capture",
                args=(self.config, self.feature_metrics, publisher),
            )
```

/events/audio.py
```
def detect_audio(self, audio, publisher) -> None:
    ...
    publisher.publish({"dBFS": dBFS, "rms": rms})
    ...
```

**Comment by NickM-27** [comment]
Not quite, this new class would be a `Communicator` not a dispatcher. It would be created with a queue and the subscribe would hold the queue reading logic. The `publish` would actually be a no-op

**Comment by NickM-27** [comment]
also the queue would need to have the topic, payload, and retain passed through it

**Comment by jpverdejo** [comment]
I'm kind of lost. If this will be a separate communicator and publish will be no-op, how is the audio process publishing the data so it's dispatched to all the configured dispatchers (ws and mqtt)?

**Comment by NickM-27** [comment]
You have the arch wrong. WS and MQTT are other communicators, there is only one dispatcher.

**Comment by NickM-27** [comment]
I think in general looks good except it shouldn't be so specific. I think we just want it to be `InterProcessCommunicator` or something so that it can be used for other things as well.

File: frigate/events/audio.py [hunk 1]
```diff
@@ -168,9 +174,21 @@ def detect_audio(self, audio) -> None:
...
+        # Calculate RMS (Root-Mean-Square) which represents the average signal amplitude
+        # Note: np.float32 isn't serializable, we must use np.float64 to publish the message
+        rms = np.sqrt(np.mean(np.absolute(audio_as_float**2)))
+
+        # Transform RMS to dBFS (decibels relative to full scale)
+        dBFS = 20 * np.log10(np.abs(rms) / AUDIO_MAX_BIT_RANGE)
+
+        self.stream_metadata_communicator.queue.put(
+            (self.config.name, {"dBFS": float(dBFS), "rms": float(rms)})
+        )
```
Comment (NickM-27) [inline frigate/events/audio.py:190]: Might be good to make this publishing a separate function that just gets the audio_as_float passed in

Comment (jpverdejo) [reply frigate/events/audio.py]: Makes sense. Fixed

File: frigate/events/audio.py [hunk 2]
```diff
@@ -179,6 +188,21 @@ def detect_audio(self, audio) -> None:
...
+
+        # Transform RMS to dBFS (decibels relative to full scale)
+        dBFS = 20 * np.log10(np.abs(rms) / AUDIO_MAX_BIT_RANGE)
+
+        self.inter_process_communicator.queue.put(
+            (f"{self.config.name}/metadata/dBFS", float(dBFS))
+        )
+        self.inter_process_communicator.queue.put(
+            (f"{self.config.name}/metadata/rms", float(rms))
+        )
```
Comment (NickM-27) [inline frigate/events/audio.py:204]:
```suggestion
        self.inter_process_communicator.queue.put(
            (f"{self.config.name}/audio/dBFS", float(dBFS))
        )
        self.inter_process_communicator.queue.put(
            (f"{self.config.name}/audio/rms", float(rms))
        )
```

File: frigate/comms/dispatcher.py [hunk 3]
```diff
@@ -83,6 +83,9 @@ def _receive(self, topic: str, payload: str) -> None:
                 return
         elif topic == "restart":
             restart_frigate()
+        elif "metadata" in topic:
+            # example /cam_name/metadata/dbfs payload=-55.23
+            self.publish(topic, payload, retain=True)
```
Comment (NickM-27) [inline frigate/comms/dispatcher.py:88]:
```suggestion
        else:
            self.publish(topic, payload, retain=False)
```

Comment (NickM-27) [reply frigate/comms/dispatcher.py:88]: We don't want to retain these messages

File: frigate/comms/inter_process.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+from frigate.comms.dispatcher import Communicator
+
+
+class InterProcessCommunicator(Communicator):
+    def __init__(self, queue: Queue) -> None:
+        self.queue = queue
+        self.stop_event: MpEvent = mp.Event()
+
+    def publish(self, topic: str, payload: str, retain: bool) -> None:
+        pass
```
Comment (NickM-27) [inline frigate/comms/inter_process.py:18]:
```suggestion
    def publish(self, topic: str, payload: str, retain: bool) -> None:
        """There is no communication back to the processes."""
        pass
```

**Review by NickM-27 — CHANGES_REQUESTED** [review]
I think this is better fit under the existing `audio` topic for each camera. Also, I think it should be safe to just publish any received messages that don't match the known special types.

**Review by NickM-27 — APPROVED** [review]

File: frigate/app.py [hunk 5]
```diff
@@ -205,6 +206,9 @@ def init_queues(self) -> None:
         # Queue for timeline events
         self.timeline_queue: Queue = mp.Queue()
 
+        # Queue for inter process communication
+        self.inter_process_queue: Queue = mp.Queue()
```
Comment (blakeblackshear) [inline frigate/app.py:210]: This should be updated to use the faster queues that were recently merged.

File: frigate/events/audio.py [hunk 6]
```diff
@@ -168,9 +174,12 @@ def detect_audio(self, audio) -> None:
         if not self.feature_metrics[self.config.name]["audio_enabled"].value:
             return
 
-        waveform = (audio / AUDIO_MAX_BIT_RANGE).astype(np.float32)
+        audio_as_float = audio.astype(np.float32)
+        waveform = audio_as_float / AUDIO_MAX_BIT_RANGE
         model_detections = self.detector.detect(waveform)
 
+        self.calculate_audio_levels(audio_as_float)
```
Comment (blakeblackshear) [inline frigate/events/audio.py:181]: At what rate is this called? Are we constantly publishing messages for these values?

Comment (NickM-27) [reply frigate/events/audio.py:181]: that's a good point, the ffmpeg command outputs an audio frame once a second which is likely too frequent for this message to be published.

File: frigate/app.py (hunk 5 above)
Comment (jpverdejo) [reply frigate/app.py]: Updated

**Comment by blakeblackshear** [comment]
Can we fixup the conflicts? I think we are good to merge this after that's resolved.

**Review by NickM-27 — CHANGES_REQUESTED** [review]

File: frigate/app.py [hunk 7]
```diff
@@ -226,6 +227,9 @@ def init_queues(self) -> None:
         # Queue for timeline events
         self.timeline_queue: Queue = ff.Queue(DEFAULT_QUEUE_BUFFER_SIZE)
 
+        # Queue for inter process communication
+        self.inter_process_queue: Queue = ff.Queue()
```
[the remark this answers is not in the document]
Comment (jpverdejo) [reply frigate/app.py]: Done

**Review by NickM-27 — APPROVED** [review]

**Pull request merged**